### PR TITLE
Enumerable methods

### PR DESCRIPTION
**Pull request by hillarioh**
This is a solo project in the basic ruby course that entails creating methods that perform the same function as enumerable methods.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,45 @@
-# ruby-enumerable
\ No newline at end of file
+# ruby-enumerable
+
+It is a project about building methods with same functionality as the enurable methods
```
Comment (mosesogwo) [inline README.md]: You could include a list of the methods created.

**Review by mosesogwo — CHANGES_REQUESTED** [review]
**Hi @hillarioh, good work:+1: but we need to fix the following:**
- For **my_each, my_each_with_index , my_select and my_map** methods, they should return an enumerator with members of the given enum when no block is given instead of returning `nil`

- For **my_all?, my_any? and my_none?** methods, make sure you handle the scenarios when a Class, or a Regex or any other pattern is given as argument. See this [documentation](https://ruby-doc.org/core-2.6.5/Enumerable.html) for the expected behaviours. The methods should be able to receive an optional argument so that you can have operations like `['a', 'b', 'c'].my_all?(String)`, to return `true`

- **my_all?** should return true when no block is given and none of the members is false or nil.

- **my_any?** should return true when no block is given and there is at least one member that is not false or nil.

- **my_none?** should return true when no block is given and none of the members of the given enum is true.

- **my_inject** should be able to take a symbol as argument and combine the elements of the given enum using the named binary operation of that symbol. Example `[1, 2, 3, 5].my_inject(:+)` should return `11`. IT should also be able to take a starting value and symbol as second argument.


_**Kindly make these changes and submit another code review request.**_

**Any questions? You can reach me on Slack (Moses Ogwo)**

**Cheers:stars:**

### Status: Changes Requested.

File: script.rb [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    return puts "#{self} has no blockk" unless block_given?
+
+    while i < size
+      yield(self[i])
+      i += 1
+    end
+  end
+
+  def my_each_with_index
+    return puts "#{self} has no blockk" unless block_given?
```
Comment (Forison) [inline script.rb]: #### return enum(:name_of method) unless block_given?

File: script.rb [hunk 3]
```diff
@@ -0,0 +1,230 @@
+## by Hillarioh
+
+module Enumerable
+  def my_each
+    i = 0
+
+    return puts "#{self} has no blockk" unless block_given?
```
Comment (Forison) [inline script.rb]: #### return enum(:name_of method) unless block_given?

File: script.rb [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+      yield(self[i], i)
+
+      i += 1
+    end
+  end
+
+  def my_select
+    arrayed = []
+
+    return puts "#{self} has no block" unless block_given?
```
Comment (Forison) [inline script.rb]: return enum(:name_of method) unless block_given?

**Review by Forison — CHANGES_REQUESTED** [review]
## Status:: Require changes❌ 
# Hi @hillarioh,
You have done a good job so far, the issues raised in the previous review have not been addressed🤔. Here are a few changes you have to make to get this project approved.
> note: your methods are supposed to mimic the behavior(functionalities) of ruby's preexisting enumerable methods.
* [ ] ```my_each```, ```my_each_with_index```, ```my_select``` and ```my_map``` ought to return an enumerator when no block is given.
* [ ] As mentioned earlier ```p ['a', 'b', 'c'].my_all?(String)``` should return ```true```
* [ ] As mentioned earlier, my_all? should return true when no block is given and none of the members is false
* [ ] As mentioned earlier, my_any? should return true when no block is given and there is at least one member that is not false or nil.
* [ ] As mentioned earlier, my_none? should return true when no block is given and none of the members of the given enum is true.
* [ ] As mentioned earlier, my_inject should be able to take a symbol as argument and combine the elements of the given enum using the named binary operation of that symbol. Example [1, 2, 3, 5].my_inject(:+) should return 11. IT should also be able to take a starting value and symbol as second argument.
## check my comments below👇
## That is it, kindly make these changes and submit the project for review😊.
#### Happy coding 😎
Addo

**Review by pelzolga123 — CHANGES_REQUESTED** [review]
The project requires changes.

**Comment by pelzolga123** [comment]
## Hi Lecrae!  
### You did a good job so far!
But still, not all methods work correctly:
*  #### My_all?
* [ ] When a pattern other than Regex or a Class is given should return true if all of the collection matches the pattern. For example:
`[3,3,3].my_all?(3)` - should return true

*  #### My_any?
* [ ] When no block or argument is given should return true if at least one of the collection is not false or nil. For example:
`[nil, false, true, []].my_any?` - should return true

* [ ] When a pattern other than Regex or a Class is given should return false if none of the collection matches the pattern. For example:
`['dog', 'bird', 'fish'].my_any?('cat')` - should return false

*  #### My_none?
* [ ] When no block or argument is given should return true only if none of the collection members is true. For example:
`[nil, false, true, []].my_none?` - should return false

* [ ] When a pattern other than Regex or a Class is given should return true only if none of the collection matches the pattern. For example:
`['dog', 'bird', 'fish'].my_none?(5)` - should return true

*  #### My_inject
* [ ] Should combine all elements of enum by applying a binary operation, specified by a block. To test your code you can use an example below, `my_inject` method should return the same value as `inject`.
```ruby
array = Array.new(10) { rand(0...10) }
operation = proc { |sum, n| sum + n }

p array.my_inject(&operation) 
p array.inject(&operation)
```
 
* [ ] When a symbol is specified should combine each element of the collection by applying the symbol as a named method. To test your code you can use an example below, `my_inject` method should return the same value as `inject`.
```ruby
array = Array.new(10) { rand(0...10) }

p array.my_inject(:+) 
p array.inject(:+)
```
### Please kindly fix the above issues and send another code review request.

If you have any questions or need help, you can message me on slack (@ Olga Pelts). *Happy coding*

File: script.rb [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+    return true if size.zero?
+
+    arrayed = []
+
+    if val.class == Regexp
+      my_each { |i| arrayed << val.match?(i) }
+    elsif block_given?
+      my_each { |i| arrayed << yield(i) }
+
+    elsif !val.nil? && !(val.instance_of? Integer) && !(val.instance_of? String)
```
Comment (shshamim63) [inline script.rb]: not only Integer or String `val` can be an instance of Array to. So covering them using a specific class, gonna be really long. think about using a .class. You can check `puts 3.class` on a ruby compiler. And can modify that to reach your goal.

File: script.rb [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+    while j < arrayed.size
+      if arrayed[j] != true
+        state = false
+        break
+      end
+      j += 1
+    end
+
+    state
+  end
```
Comment (shshamim63) [inline script.rb]: `my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)) 
- when a class is passed as an argument returns true if all of the collection is a member of such class
```ruby
[1, 2, 3].my_all?(Integer) #should return true
[1, 2, nil].my_all?(Integer) #should return false
[[1, 2, 3],[1, 2]].my_all?(Array)  #should return true
```
**There are a lot of options to refactor the code. Try to find out the core conditions. After that figure out the sub supporting condition. The core condition I can see is if a block is given or not.**

File: script.rb [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+    while j < arrayed.size
+      if arrayed[j] == true
+        state = true
+        break
+      end
+      j += 1
+    end
+
+    state
+  end
```
Comment (shshamim63) [inline script.rb]: `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- when no block or argument is given returns true if at least one of the collection is not false or nil
```ruby
[1, 2, 3].my_any? #should return true
[1, 2, nil].my_any? #should return true
[false, false, nil].my_any? #should return false
```
- when a class is passed as an argument returns true if at least one of the collection is a member of such class
```ruby
[1, 2, 3].my_any?(Integer) #should return true
[1, 'demo', false].my_any?(Integer) #should return true
['demo', false, nil].my_any?(Integer) #should return false
```

File: script.rb [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+        state = false
+        break
+
+      end
+      j += 1
+
+    end
+
+    state
+  end
```
Comment (shshamim63) [inline script.rb]: `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- when a class is passed as an argument returns true if none of the collection is a member of such class
```ruby
[1, 2, 3].my_none?(String)  #should return true
['1', '2', '3'].my_none?(String)  #should return false
```

File: script.rb [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+    i = 0
+
+    while i < size - 1
+      self[i + 1] = yield(self[i], self[i + 1]) if block_given?
+      result = self[i + 1] if i == size - 2
+      i += 1
+    end
+
+    result
+  end
```
Comment (shshamim63) [inline script.rb]: `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- combines all elements of enum by applying a binary operation, specified by a block
```ruby
[1, 2, 3].my_inject { |memo, num| memo + num } #should return 6
[1, 2, 3].my_inject(1) { |memo, num| memo + num } #should return 7
```
- when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
[1, 2, 3].my_inject(:+)  #should return 6
[1, 2, 3].my_inject(1, :+)  #should return 7
```
- inject can perform on range too.
___
Well, you can say that you are having the exact result. But your approach is updating the existing array, it should not. Take a look at the following code. And try to execute it in your main.rb
```ruby
ARRAY_SIZE = 10
LOWEST_VALUE = 1
HIGHEST_VALUE = 20
array = Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) }
operation = proc { |sum, n| sum + n }

p array.inject(&operation)
p array.my_inject(&operation)

print array
p array.my_inject(:+)
print array
p array.inject(:+)

p array.my_inject(3, :*) == array.inject(3, :*)
p array.inject(3, :*) == array.my_inject(3, :*)
```

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## Status :traffic_light: Required Changes :tractor: 
___
Hi @hillarioh :wave: 
Good job :clap: , you are almost done. Let's cover the addressed issues to declare it as complete.
___
### :spades:  Mandatory Changes :spades: 
___
1. I can see that you have altered the rubocop config file. You should not do that. Please make sure that you are using [this](https://github.com/microverseinc/linters-config/blob/master/ruby/.rubocop.yml) config.
2. Please fix the methods below by following the comments attached sections.
- my_all?
- my_none?
- my_any?
- my_inject
____
##### Please fix the mentioned issues and send another review request.
____
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

File: script.rb [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+    while j < arrayed.size
+      if arrayed[j] == true
+        state = true
+        break
+      end
+      j += 1
+    end
+
+    state
+  end
```
Comment (stratospherique) [inline script.rb]: * [ ] should have working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class
  ```ruby
    array = [1,2,5,4,7]
    p array.my_any?(Integer) == array.any?(Integer) # false
  ```

File: script.rb [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+        state = false
+        break
+
+      end
+      j += 1
+
+    end
+
+    state
+  end
```
Comment (stratospherique) [inline script.rb]: * [ ] should have working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

* [ ] my_none? when no block or argument is given returns true only if none of the collection members is true
  ```ruby
    array = [nil, false, nil, false]
    p array.my_none? == array.none? # false
  ```

File: script.rb [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+# p [1, 2, 3].my_any? # should return true
+# p [1, 2, nil].my_any? # should return true
+# p [false, false, nil].my_any? # should return false
+# p [[], []].my_any? # should return true
+
+# TEST CASES - NONE
+p [nil, false, true, []].my_none? # should return false
+p %w[dog bird fish].my_none?(5) # should return true
+p [1, 2, 3].my_none?(String) # should return true
+p %w[1 2 3].my_none?(String) # should return false
```
Comment (stratospherique) [inline script.rb]: - **[REQUIRED]** All commented or debugging related code should not be committed. :pray:  delete it.

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Impressive work @hillarioh :clap: :clap: However, few methods need further refactoring in order to get approval for this project. Check :arrow_down: 

- **[REQUIRED]** Check the code comments below, they should help you refactor your methods :gear: 

### That's it :relaxed: When finished, kindly submit another code re-review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

File: script.rb [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+        state = false
+        break
+
+      end
+      j += 1
+
+    end
+
+    state
+  end
```
Comment (stratospherique) [inline script.rb]: * [ ] my_none? when no block or argument is given returns true only if none of the collection members is true
  ```ruby
    array = [1,2,5,4,7]
    p array.my_none? == array.none? # false
  ```

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great work implementing the suggested changes :clap: However, one method needs a bit of refactoring in order to get approval for this project. Check :arrow_down: 

- **[REQUIRED]** Check the code comments below, they should help you refactor your methods

### That's it :relieved: When finished, kindly submit another code re-review request :pray: 

Happy coding :v: 
**_Ahmed Mahfoudh_**

File: README.md [hunk 14]
```diff
@@ -1 +1,55 @@
...
+- #my_inject
+
+# What it does
+
+Uses loops and other syntax to provide the same output as if using the original enumerable methods
+
+## Demo
+
+This is a link to demo
+Repl.it(https://repl.it/@okerioh/enumarable-methods )
```
Comment (mosesogwo) [inline README.md:22]: You can make this a hyperlink by writing this: `[Repl.it](https://repl.it/@okerioh/enumarable-methods)`

**Review by mosesogwo — APPROVED** [review]
**Congratulations:trophy: I applaud your doggedness:+1:**

You can now merge this branch.

**Cheers:stars:**

### Status: Approved:heavy_check_mark:

**Pull request merged**